lib/jwt/jwk/ec.rb

```diff
@@ -11,37 +11,44 @@ class EC < KeyBase # rubocop:disable Metrics/ClassLength
       KTY    = 'EC'
       KTYS   = [KTY, OpenSSL::PKey::EC].freeze
       BINARY = 2
+      EC_PUBLIC_KEY_ELEMENTS = %i[kty crv x y].freeze
+      EC_PRIVATE_KEY_ELEMENTS = %i[d].freeze
+      EC_KEY_ELEMENTS = (EC_PRIVATE_KEY_ELEMENTS + EC_PUBLIC_KEY_ELEMENTS).freeze
 
-      attr_reader :keypair
+      def initialize(keypair, params = nil, options = {})
+        params ||= {}
 
-      def initialize(keypair, options = {})
-        raise ArgumentError, 'keypair must be of type OpenSSL::PKey::EC' unless keypair.is_a?(OpenSSL::PKey::EC)
+        # For backwards compatibility when kid was a String
+        params = { kid: params } if params.is_a?(String)
 
-        @keypair = keypair
+        # Accept OpenSSL key as input
+        keypair = parse_ec_key(keypair) if keypair.is_a?(OpenSSL::PKey::EC)
 
-        super(options)
+        raise ArgumentError, 'keypair must be of type OpenSSL::PKey::EC' unless keypair.is_a?(Hash)
```

Comment:
I have a few suggestions for the initializer (the suggestion applies for all the different types).
 
- The `keypair` parameter could be renamed to something more describing.
- The given OpenSSL object could be kept to avoid recreation from JWK parameters
- The ArgumentError could tell that the parameter also can be a Hash
- Transforming the keys to symbols is only needed when first parameter is a Hash.
 
one idea would be to handle the different variants of the first parameters in a case-block:

```
def initialize(keypair_or_params, params = nil, options = {})
  params ||= {}

  # For backwards compatibility when kid was a String
  params = { kid: params } if params.is_a?(String)

  key_params = case keypair_or_params
               when OpenSSL::PKey::EC  # Accept OpenSSL key as input
                 @keypair = keypair_or_params # Preserve the object to avoid recreation
                 parse_ec_key(keypair_or_params)
               when Hash
                 keypair_or_params.transform_keys(&:to_sym)
               else
                 raise ArgumentError, 'keypair_or_params must be of type OpenSSL::PKey::EC or Hash with key parameters'
               end

  params = params.transform_keys(&:to_sym)

  check_jwk(keypair, params)

  super(options, key_params.merge(params))
end
```

Reply:
What do you think, would an example for the usage without the OpenSSL object in the readme be valuable?

Reply:
> What do you think, would an example for the usage without the OpenSSL object in the readme be valuable?

Yes. In fact, the section `Importing and exporting JSON Web Keys` should probably demonstrate this instead of converting an OpenSSL key as in the example above.

> The `keypair` parameter could be renamed to something more describing.

I agree, though `keypair_or_params` runs the risk of suggesting that you can specify everything from the first argument in the second, which is not the case (e.g. the first argument needs to be the one describing the `kty`).

In some other places I think I have simply used `key`, which could be an OpenSSL Key or a JSON Web Key.

> The given OpenSSL object could be kept to avoid recreation from JWK parameters

I like this :+1: